Node authorizer

liggitt opened the pull request:
This PR implements the authorization portion of https://github.com/kubernetes/community/blob/master/contributors/design-proposals/kubelet-authorizer.md and kubernetes/features#279:
* Adds a new authorization mode (`Node`) that authorizes requests from nodes based on a graph of related pods,secrets,configmaps,pvcs, and pvs:
  * Watches pods, adds edges (secret -> pod, configmap -> pod, pvc -> pod, pod -> node)
  * Watches pvs, adds edges (secret -> pv, pv -> pvc)
  * When both Node and RBAC authorization modes are enabled, the default RBAC binding that grants the `system:node` role to the `system:nodes` group is not automatically created.
* Tightens the `NodeRestriction` admission plugin to require identifiable nodes for requests from users in the `system:nodes` group.

This authorization mode is intended to be used in combination with the `NodeRestriction` admission plugin, which limits the pods and nodes a node may modify. To enable in combination with RBAC authorization and the NodeRestriction admission plugin:
* start the API server with `--authorization-mode=Node,RBAC --admission-control=...,NodeRestriction,...`
* start kubelets with TLS boostrapping or with client credentials that place them in the `system:nodes` group with a username of `system:node:<nodeName>`

```release-note
kube-apiserver: a new authorization mode (`--authorization-mode=Node`) authorizes nodes to access secrets, configmaps, persistent volume claims and persistent volumes related to their pods.
* Nodes must use client credentials that place them in the `system:nodes` group with a username of `system:node:<nodeName>` in order to be authorized by the node authorizer (the credentials obtained by the kubelet via TLS bootstrapping satisfy these requirements)
* When used in combination with the `RBAC` authorization mode (`--authorization-mode=Node,RBAC`), the `system:node` role is no longer automatically granted to the `system:nodes` group.
```

```release-note
RBAC: the automatic binding of the `system:node` role to the `system:nodes` group is deprecated and will not be created in future releases. It is recommended that nodes be authorized using the new `Node` authorization mode instead. Installations that wish to continue giving all members of the `system:nodes` group the `system:node` role (which grants broad read access, including all secrets and configmaps) must create an installation-specific ClusterRoleBinding.
```

Follow-up:
- ~~enable e2e CI environment with admission and authorizer enabled~~ https://github.com/kubernetes/kubernetes/pull/40760 https://github.com/kubernetes/kubernetes/pull/46796
- ~~enable this authorizer and admission plugin in kubeadm~~ https://github.com/kubernetes/kubernetes/pull/46879
- ~~enable this authorizer and admission plugin in kube-up~~ https://github.com/kubernetes/kubernetes/pull/46796
- ~~scale-test in kubemark~~ https://github.com/kubernetes/kubernetes/pull/46921

liggitt committed:
Separate system:node rules

liggitt committed:
Add PV util for extracting referenced secrets

liggitt committed:
bump(golang.org/x/tools/container/intsets): 2382e3994d48b1d22acc2c86bcad0a2aff028e32

liggitt committed:
bump(github.com/gonum/graph): 50b27dea7ebbfb052dfaf91681afc6fde28d8796

ncdc commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,219 @@
+
+const (
+	ConfigMapNode nodeType = iota
+	NodeNode      nodeType = iota
```
iota copypasta?

ncdc commented on plugin/pkg/auth/authorizer/node/graph_populator.go:
```diff
@@ -0,0 +1,41 @@
+type GraphPopulator struct {
+	graph *Graph
+
+	podInformer              cache.SharedIndexInformer
```
We generally haven't been storing refs to the underlying `cache.SharedIndexInformer` objects. If you're using a shared informer, it's generally safe to assume that something else (e.g. kube-controller-manager) will start it.

ncdc commented on plugin/pkg/auth/authorizer/node/graph_populator.go:
```diff
@@ -0,0 +1,41 @@
+}
+
+func (g *GraphPopulator) Run(stopCh <-chan struct{}) {
+	g.podInformer.GetController().Run(stopCh)
```
These 2 lines are likely no-ops (see my comment above, https://github.com/kubernetes/kubernetes/blob/44127c588d24ecabe3db7db66828b181f197a3c7/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L261-L263, and https://github.com/kubernetes/kubernetes/blob/44127c588d24ecabe3db7db66828b181f197a3c7/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L161).

ncdc commented on plugin/pkg/auth/authorizer/node/graph_populator_pod.go:
```diff
@@ -0,0 +1,37 @@
+	}
+	if oldObj != nil && pod.Spec.NodeName == oldObj.(*api.Pod).Spec.NodeName {
+		// Node is unchanged
+		// TODO: do we need to check for uid changes?
```
I would say yes

liggitt commented:
Comments addressed. Still working on tests, and I'll have some follow-ups for performance/memory related improvements to the graph, but the authorizer and graph population code is ready for review

kargakis commented on third_party/forked/gonum/graph/README.md:
```diff
@@ -0,0 +1 @@
+Forked from gonum/graph@50b27dea7ebbfb052dfaf91681afc6fde28d8796 to support memory-use improvements to the simple graph
```
Any plans to upstream the changes?

liggitt replied:
Absolutely, I just don't want to block on the upstream PRs being accepted and vendorable prior to code freeze

liggitt committed:
gonum: slice-based edge holder

Use slices to store sets of edges for n <= 4.
~20% memory savings with many nodes with n=2 edges.

liggitt committed:
gonum: directed acyclic graph

Implements graph.Directed capable of storing at most one edge between any two nodes.
Uses the Undirected implementation for space efficiency (~30% space savings).

liggitt committed:
gonum: visiting graph traversal

Use visitors to avoid allocating slices to hold nodes during traversal (and to allow short-circuiting)
Benchmarked at 95% space savings traversing nodes with many edges.

liggitt commented:
@k8s-bot bazel test this
@k8s-bot verify test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot kops aws e2e test this
@k8s-bot pull-kubernetes-federation-e2e-gce test this
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

deads2k commented on hack/local-up-cluster.sh:
```diff
@@ -197,6 +197,7 @@ KUBELET_HOST=${KUBELET_HOST:-"127.0.0.1"}
 API_CORS_ALLOWED_ORIGINS=${API_CORS_ALLOWED_ORIGINS:-/127.0.0.1(:[0-9]+)?$,/localhost(:[0-9]+)?$}
 KUBELET_PORT=${KUBELET_PORT:-10250}
 LOG_LEVEL=${LOG_LEVEL:-3}
+LOG_SPEC=${LOG_SPEC:-""}
```
give me a comment with an example of how to set this usefully?

deads2k commented on pkg/registry/rbac/rest/storage_rbac.go:
```diff
@@ -176,7 +178,8 @@ func PostStartHook(hookContext genericapiserver.PostStartHookContext) error {
 		// ensure bootstrap rolebindings are created or reconciled
-		for _, clusterRoleBinding := range append(bootstrappolicy.ClusterRoleBindings(), bootstrappolicy.ControllerRoleBindings()...) {
+		bindNodeRole := !utilfeature.DefaultFeatureGate.Enabled(features.NodeAuthorizer)
+		for _, clusterRoleBinding := range append(bootstrappolicy.ClusterRoleBindings(bindNodeRole), bootstrappolicy.ControllerRoleBindings()...) {
```
Yuck.  Applying filters may be more palatable since it won't lead to flag madness and the role names have to stay pretty stable. No option is pretty.

liggitt replied:
I'd like to do the following:
* make enabling the node authorizer skip this binding (implemented in this PR)
* announce the default `system:nodes` binding as deprecated now, and remove it completely in 1.8/1.9 timeframe, at which point we can get rid of this flag.

liggitt replied:
switched to a filter that gets enabled when using the node authorizer. it's ugly, but will go away in 1.8. try not to fixate on it.

deads2k commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+	if len(n.namespace) == 0 {
+		return nodeTypes[n.nodeType] + ":" + n.name
+	}
+	return nodeTypes[n.nodeType] + ":" + n.namespace + "/" + n.name
```
nit: why doesn't this match our CLI?

liggitt replied:
`type/name`? how would you fit in the namespace?

deads2k commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+
+// NodeEdge is a graph edge that is related to a kubernetes Node
+type NodeEdge struct {
+	F    graph.Node
```
`From` and `To` please.

liggitt replied:
conflicts with method names in interface

deads2k commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+type NodeEdge struct {
+	F    graph.Node
+	T    graph.Node
+	Node graph.Node
```
This looks bonkers, but I'm guessing you intend for this to be a `KubernetesNode` or maybe a `RootNode` or its pointing  to some important node that you want a direct link to in your edge?  Regardless, see an Edge with a Node embedded in it looks really strange, find a better name.

deads2k commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+type Graph struct {
+	lock  sync.RWMutex
+	graph *simple.DirectedAcyclicGraph
+	nodes map[nodeType]map[string]map[string]*NamedNode
```
godoc.  map of node type to map of ....something (blind guess, namespace?)... to map of ....something (blind guess, name?).... to graph node

timstclair replied:
I suggest creating a key type instead (see [https://blog.golang.org/go-maps-in-action](https://blog.golang.org/go-maps-in-action#TOC_5.))

liggitt replied:
a flat map, keyed by a multi-field key type bumped memory allocations during population by 15%, and increased retained memory in a populated graph by 15%

timstclair replied:
Wow, that's surprising to me. It would be interesting to do a write up on that...

deads2k commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+type nodeType byte
+
+const (
+	ConfigMapNode nodeType = iota
```
Comment defending this.  Size I assume.

liggitt replied:
yes

deads2k commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+
+const (
+	ConfigMapNode nodeType = iota
+	NodeNode
```
***really?!***

ericchiang replied:
Since this package is already called `node` can't we just call this `TypeNode`? `node.TypeNode`, `node.TypePod` vs. `node.NodeNode`, `node.PodNode`.

ericchiang replied:
Since this package is already called `node` can't we just call this `TypeNode`? `node.TypeNode`, `node.TypePod` vs. `node.NodeNode`, `node.PodNode`.

liggitt replied:
I'm sorry for everything I've ever done. renaming to vertex helped

deads2k commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+}
+
+// must be called under a write lock
+func (g *Graph) getOrCreateNode(nodeType nodeType, namespace, name string) *NamedNode {
```
getOrCreateNode_locked

deads2k commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+}
+
+// must be called under a read lock
+func (g *Graph) getNode(nodeType nodeType, namespace, name string) (*NamedNode, bool) {
```
getNode_rlocked

deads2k commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+		g.nodes[nodeType] = typedNodes
+	}
+
+	namespacedNodes, exists := typedNodes[namespace]
```
I'm going nuts.  Its inconsistent with gonum graph, but I'm considering `Vertex`

liggitt replied:
that helped

deads2k commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+
+	// find existing neighbors
+	neighborsToRemove := []graph.Node{}
+	g.graph.VisitFrom(node, func(neighbor graph.Node) bool {
```
I don't recognize this method from the library.

deads2k replied:
Ok, I think this is collecting a list of downstream vertices which have no other edges.  If that's correct, please comment it.  If that's incorrect, definitely comment it.

liggitt replied:
my addition (see `gonum: visiting graph traversal` commit) to save allocating a slice of all neighbors for every step of every read traversal (we really need reads to be cheap and fast). saves ~95% on allocations for highly shared secrets

deads2k commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+	if len(g.nodes[nodeType][namespace]) == 0 {
+		delete(g.nodes[nodeType], namespace)
+	}
+	if len(g.nodes[nodeType]) == 0 {
```
this looks unnecessary.  number of different nodetypes is minimal and they will likely reappear.  I see no reason to remove this.

deads2k commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+	g.deleteNode(PodNode, pod.Namespace, pod.Name)
+	podNode := g.getOrCreateNode(PodNode, pod.Namespace, pod.Name)
+	nodeNode := g.getOrCreateNode(NodeNode, "", pod.Spec.NodeName)
+	g.graph.SetEdge(NewNodeEdge(podNode, nodeNode, nodeNode))
```
I'm assuming this is a guiding principle.  All edges point towards the KubernetesNodeVertex.  If so, doc the overall structure.

deads2k commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+//   configmap -> pod
+//   pvc -> pod
+//   pod -> node
+func (g *Graph) AddPod(pod *api.Pod) {
```
Only valid to be called if `pod.Spec.NodeName` has been set?  Is that field immutable after its been set?

liggitt replied:
yes

deads2k commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+		return true
+	})
+
+	for _, v := range pod.Spec.Volumes {
```
visitor, visitor, goose!

deads2k commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+	pvNode := g.getOrCreateNode(PVNode, "", pv.Name)
+
+	if pv.Spec.ClaimRef != nil {
+		g.graph.SetEdge(simple.Edge{F: pvNode, T: g.getOrCreateNode(PVCNode, pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name)})
```
Why is this edge different?  It's not immediately obvious to me.  I'm assuming its related to the fact that the PV doesn't directly reference a kubernetesNode?  The asymmetry takes more thought than I'd like.  Comment please.

liggitt replied:
right, we don't know the pvc->pod->node info from here (and that part of the graph may not even exist yet). commented

deads2k commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+	defer g.lock.Unlock()
+
+	g.deleteNode(PVNode, "", pv.Name)
+	pvNode := g.getOrCreateNode(PVNode, "", pv.Name)
```
Why would I add this if it had no edges?

deads2k replied:
> Why would I add this if it had no edges?

If it had no shot at edges.

liggitt replied:
moved the re-add inside the pvc check

deads2k commented on plugin/pkg/auth/authorizer/node/node_authorizer.go:
```diff
@@ -0,0 +1,169 @@
+		return false, "unknown node", nil
+	}
+
+	if !rbac.RulesAllow(attrs, r.nodeRules...) {
```
This check is actually faster?

liggitt replied:
moved it after the subdivided checks

deads2k commented on plugin/pkg/auth/authorizer/node/node_authorizer.go:
```diff
@@ -0,0 +1,169 @@
+	}
+
+	if len(nodeName) == 0 {
+		// allow requests from unidentifiable nodes if not in strict mode
```
This doesn't match my expectations.  It breaks in cases where RBAC is disabled and a later authorizer would not have allowed access.  You should return false and allow the next authorizer in the chain to authorize.

deads2k replied:
This bit *still* doesn't match my expectations.  If you don't want it to be strict, it seems incumbent upon you (the cluster-admin) to provide a different rule which will allow.

liggitt replied:
> Do we really care about supporting that use case here? Or can we just recommend that the node authorizer not be enabled unless all nodes are identifiable (e.g. use the legacy RBAC rules otherwise).

I think I agree, and actually think it would be better to make this a normal addressable authorizer, documented as requiring conforming node identities to use. Same for the node admission plugin... always run in strict mode, deployments can omit it if they have nodes that don't identify themselves correctly.

Recommended config would end up looking like `--authorization-mode=Node,RBAC --admission-control=...,NodeRestriction,...`

deads2k commented on plugin/pkg/auth/authorizer/node/node_authorizer.go:
```diff
@@ -0,0 +1,169 @@
+	if attrs.GetAPIGroup() != api.GroupName {
+		// the resources we care about subdividing happen to be in the core API group.
+		// access to resources in other API groups is not subdivided.
+		return true, "", nil
```
This is again, RBAC centric.  I understand why you want to have a (potentially) faster precheck with RBAC rules, but I don't think that this authorizer can reasonably grant access based on rules that are not necessarily active in the process.

deads2k commented on plugin/pkg/auth/authorizer/node/node_authorizer.go:
```diff
@@ -0,0 +1,169 @@
+
+	// subdivide access to specific resources
+
+	if attrs.GetAPIGroup() != api.GroupName {
```
The list of `GroupResources` that you're checking is a clear whitelist.  I'd prefer a tight existence check. Since you'll want zero allocations I'm fine with an exploded if/else structure.

deads2k commented on plugin/pkg/auth/authorizer/node/node_authorizer.go:
```diff
@@ -0,0 +1,169 @@
+
+	default:
+		// Access to other resources is not subdivided
+		return true, "", nil
```
Alright, I've gotten there.  If you have no other authorizer for a node, you still need this authorizer to allow you through.  Tidy the white list, add a better comment, and this can fail closed to make it more obvious to slow readers like me.

deads2k commented on plugin/pkg/auth/authorizer/node/node_authorizer.go:
```diff
@@ -0,0 +1,169 @@
+			return false, "can only get individual persistentvolumes", nil
+		}
+		return r.hasPathFrom(nodeName, PVNode, "", attrs.GetName()), "", nil
+
```
Add a block allowing access to any node with the explanation that the kubelet can already list all pods.

liggitt replied:
not sure I understand this comment. I actually moved the system:node rule check to the end, so now the function looks like:

1. reject requests that aren't from nodes
2. specially handle requests for secret/configmap/pvc/pv objects
3. for other resources, defer to the system:node rule evaluation

deads2k commented on plugin/pkg/auth/authorizer/node/node_authorizer.go:
```diff
@@ -0,0 +1,169 @@
+				if nodeEdge.NodeID() == nodeNode.ID() {
+					// We found an edge leading to the node we want
+					found = true
+					return true
```
If you found a matching edge, why are you continuing traversal?  The return doesn't do what I think it does?

liggitt replied:
right, this return just decides whether to follow this particular edge. the visitor passed to `.Walk()` is what gets to short-circuit

deads2k commented:
Simple and straightforward.  The names are really unfortunate.  I would seriously consider calling them Vertices to avoid the crazy of "is it a node or a node?"

I've got a few comments I'd like in a separate commit and I'll come back to the upstream carries.

ericchiang commented on pkg/api/persistentvolume/util.go:
```diff
@@ -0,0 +1,55 @@
+// VisitPVSecretNames invokes the visitor function with the name of every secret
+// referenced by the PV spec. If visitor returns false, visiting is short-circuited.
+// Returns true if visiting completed, false if visiting was short-circuited.
+func VisitPVSecretNames(pv *api.PersistentVolume, visitor func(string) bool) bool {
```
Why is this a visitor? I don't see any recursive calls.

liggitt replied:
no allocations of slices to hold secret names

timstclair commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+			neighborsToRemove = append(neighborsToRemove, neighbor)
+		}
+		return true
+	})
```
I'm confused... when do nodes have outgoing edges?

timstclair commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+			neighborsToRemove = append(neighborsToRemove, neighbor)
+		}
+		return true
+	})
```
If I'm following this correctly (big if), `graph.Degree` doesn't pay attention to edge directionality, so in a case like node <- pod <- secret, the pod would have degree 2, and wouldn't be removed.  Also, in the case with something (A) that is referenced by something else (B) that is referenced by 2 nodes, (A) will have degree 1, but shouldn't be removed.

I think what you might want to do is add all the visited nodes to a list, and then iteratively remove items from the list with no outgoing edges (until either the list is empty or no items are removed)

liggitt replied:
correct, degree counts incoming and outgoing edges. this is avoiding leaving orphan vertices with no edges. since we ignore any neighbor with a degree higher than 1, we leave them alone if they are connected to anyone else in addition to us.

timstclair commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,244 @@
+		if len(g.nodes[n.nodeType]) == 0 {
+			delete(g.nodes, n.nodeType)
+		}
+	}
```
Actually, why are you removing neighbors at all? If they still exist in the Kubernetes API, they should still exist in the graph, right?

liggitt replied:
no, we're only watching pods and pvs to populate this graph, and inferring the existence of configmaps/secrets, pvcs, and nodes by the references from the pods and pvs. if I remove a pod, I want to remove any secrets it referenced if nothing else is referencing them.

liggitt commented on pkg/api/persistentvolume/util.go:
```diff
@@ -0,0 +1,55 @@
+	switch {
+	case source.AzureFile != nil:
+		if len(source.AzureFile.SecretName) > 0 && !visitor(source.AzureFile.SecretName) {
+			return false
```
this is modeled after the pod method, which has other references to secrets (in containers, etc). if a PV gains those (for example, pv->storageclass->secret), it wouldn't be correct to return after just looking at the volumesource.

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -86,6 +86,50 @@ func addClusterRoleBindingLabel(rolebindings []rbac.ClusterRoleBinding) {
 }
 
+func NodeRules() []rbac.PolicyRule {
+	return []rbac.PolicyRule{
```
no changes, straight move

liggitt commented on plugin/pkg/auth/authorizer/node/node_authorizer_test.go:
```diff
@@ -0,0 +1,378 @@
+		},
+		{
+			name:   "allowed pvc",
+			attrs:  authorizer.AttributesRecord{User: node0, ResourceRequest: true, Verb: "get", Resource: "persistentvolumeclaims", Name: "pvc0-pod0-node0", Namespace: "ns0"},
```
sure... the names try to represent the objects connected by the graph with the node name at the end, so node0 should not be allowed to get any object ending in "node1"

liggitt commented on plugin/pkg/auth/authorizer/node/node_authorizer_test.go:
```diff
@@ -0,0 +1,378 @@
+		{
+			name:   "disallowed secret via pod",
+			attrs:  authorizer.AttributesRecord{User: node0, ResourceRequest: true, Verb: "get", Resource: "secrets", Name: "secret0-pod0-node1", Namespace: "ns0"},
+			expect: false,
```
the populated test tests this, but I'll add a more inspectable test as well

liggitt commented:
comments addressed in additional commits

liggitt committed:
Make NodeRestriction admission require identifiable nodes

liggitt commented:
integration test added, and authorizer and admission plugin made strict. PTAL

liggitt commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

liggitt commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

liggitt commented:
@kubernetes/sig-auth-pr-reviews

deads2k commented:
lgtm
/approve

liggitt commented:
comments addresses, rebased and squashed. @timstclair @ericchiang any other comments?

smarterclayton commented:
Fix up the comments on node roles in follow up

timstclair commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,258 @@
+	return vertexTypes[n.vertexType] + ":" + n.namespace + "/" + n.name
+}
+
+// destinationEdge is a graph edge that includes a denormalized reference to the destination node
```
nit: Add a comment to the effect that the denormalized destination is only relevant when there is a single leaf vertex reachable from the source vertex.
nit: s/node/vertex/

timstclair commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,258 @@
+}
+
+// must be called under write lock
+func (g *Graph) deleteVertex(vertexType vertexType, namespace, name string) {
```
nit: deleteVertex_locked, for consistency

timstclair reviewed (commented):
Mostly just nits. The change from node to vertex really helped.

smarterclayton commented:
Re alignment depends on whether the compiler padded you.  Generally you pack structs with "aligned, aligned, aligned, next largest, smallest" (ie ptr, ptr, int32, byte) should be more efficient.  Would have to look at the layout generated by the compiler tho.

smarterclayton commented on plugin/pkg/auth/authorizer/node/graph.go:
```diff
@@ -0,0 +1,258 @@
+	vertexType vertexType
+	name       string
+	namespace  string
+	id         int
```
http://golang-sizeof.tips/?t=Ly8gU2FtcGxlIGNvZGUKc3RydWN0IHsKCWEgc3RyaW5nCgljIHN0cmluZwoJaSBpbnQzMgogICAgYiBieXRlCgp9Cg==

is optimal - 40 bytes vs the 48 as is.  However....

http://golang-sizeof.tips/?t=Ly8gU2FtcGxlIGNvZGUKc3RydWN0IHsKICAgIGIgYnl0ZQoJYSBzdHJpbmcKCWMgc3RyaW5nCglpIGludAp9Cg==

liggitt committed:
Add Node authorization mode based on graph of node-related objects

liggitt commented:
rebased, comments addressed. will look more into graph space optimizations in 1.8 timeframe.

smarterclayton commented:
No other comments from me

/approve

Pending other reviewers sign off

timstclair commented:
LGTM, but I'll avoid adding the label to let other reviewers respond. Feel free to self apply once they have.

ericchiang commented:
/lgtm

liggitt commented:
fixed lint issue, retagging

liggitt commented:
@k8s-bot pull-kubernetes-e2e-kops-aws test this

liggitt commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this
